src/VisualStudio/Core/Def/DocumentOutline/VisualStudioCodeWindowInfoService.cs

```diff
@@ -0,0 +1,134 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.VisualStudio.Editor;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.TextManager.Interop;
+
+namespace Microsoft.VisualStudio.LanguageServices.DocumentOutline
+{
+    /// <summary>
+    /// Manages access to Visual Studio code window information in a thread-safe way.
+    /// </summary>
+    internal class VisualStudioCodeWindowInfoService
+    {
+        private readonly IVsCodeWindow _codeWindow;
+        private readonly IVsEditorAdaptersFactoryService _editorAdaptersFactoryService;
+        private readonly IThreadingContext _threadingContext;
+
+        public VisualStudioCodeWindowInfoService(
+            IVsCodeWindow codeWindow,
+            IVsEditorAdaptersFactoryService editorAdaptersFactoryService,
+            IThreadingContext threadingContext)
+        {
+            threadingContext.ThrowIfNotOnUIThread();
+            _codeWindow = codeWindow;
+            _editorAdaptersFactoryService = editorAdaptersFactoryService;
+            _threadingContext = threadingContext;
+        }
+
+        /// <summary>
+        /// Get <see cref="VisualStudioCodeWindowInfo"/>, switching to the UI thread if necessary.
+        /// </summary>
+        public async ValueTask<VisualStudioCodeWindowInfo?> GetVisualStudioCodeWindowInfoAsync(CancellationToken token)
+        {
+            if (!_threadingContext.JoinableTaskContext.IsOnMainThread)
```

Comment:
I did this because if we are already on the UI thread, with `ValueTask` nothing will be allocated, but we haven't gotten any memory pressure bugs here (yet :)) so I am happy to switch to a simpler implementation and not use `ValueTask`